### ISSUE TITLE
Merge helper apps into the tested apps (Was: TestHelper causes all apps to crash when M.exe is running)

### ISSUE THREAD

**Issue by warpech** [Starcounter/Blending#268]
Right now, TeamCity tests of the mapped Blending work, even though M.exe is not used there. For example, see this build: https://teamcity.starcounter.org/viewLog.html?buildId=62077&tab=buildResultsDiv&buildTypeId=Apps_Blending_BlendingMasterWithStarcounterDevelop24

The reason for this is a mean workaround for mapping, being:

- tests for BlendingEditor and BlendingProvider are separated. The actual integration between these BE and BP is not tested at all
- BE and BP are tested through the use of "TestHelper" apps. The TestHelper apps manipulate data used by BE and BP by having a **direct reference** to the data model between the testing app and the tested app

For the reasons listed above, our testing solution works if `M.exe` is not started. When `M.exe` is started, all apps crash during the first test!

@Starcounter/sc-mapping what do you recommend:

1. Should we write the mapping rules for the test helpers and run it with `M.exe`
2. Isn't this a bug in `M.exe` that it crashes without any notice?

Me or @kegor can show the reproduction on Hangouts.

**Comment by wanderlust-ginge** [comment]
What is the error that's being given in Dbgview when it crashes?

**Comment by warpech** [comment]
```
[1180] sc_abort(): BlendingEditor.UserSession is read-only.
```

**Comment by warpech** [comment]
## Steps to reproduce

1. In `Blending.Mapper/run.bat` replace the line https://github.com/Starcounter/Blending.Mapper/blob/b1ca816b172414a835815de64ebea93be59f039d/run.bat#L17 with:

```
star --resourcedir="%~dp0Blending\src\BlendingEditor\wwwroot" "%~dp0Blending/src/BlendingEditor/bin/debug/BlendingEditor.exe" /-api/-token=SeleniumTestToken
IF ERRORLEVEL 1 EXIT /B 1

:: Start the BlendingEditor helper apps
star "%~dp0Blending/test/BlendingEditor_TestHelper/bin/debug/BlendingEditor_TestHelper.exe"
```

2. Call `Blending.Mapper/run.bat`
3. Check that the 3 apps (M, BlendingEditor, BlendingEditor_TestHelper) are running
4. Open http://localhost:8080/blendingeditor
4. Check that all apps have crashed

This is the line that causes all apps to crash: https://github.com/Starcounter/Blending/blob/c279197375d760c24fb72f9c52cfe428ab0ab95c/test/BlendingEditor_TestHelper/Program.cs#L29

Mind that `UserSession` class is defined in `BlendingEditor`, not `BlendingEditor_TestHelper`. `BlendingEditor_TestHelper` has a reference to `BlendingEditor` in order to obtain the definition of that class.

**Comment by warpech** [comment]
For us it is important that we can test our apps. Basically I can see 2 solutions, listed in the OP.

Perhaps there is also a 3rd solution:

3. Don't do test helper apps. Instead, bake the test helper inside the tested app. There will be a challenge to do it securely.

What solution would you recommend?

I care about being productive and not seeing obscure errors like apps crashing without any information.

**Comment by warpech** [comment]
> As for your security concern, I don't see a problem. For example, you could do like @malx122 and have a special command-line flag that starts the app in "developer mode", where it serves some additional handler routes.

Ok, we will go this way.

It is not a hight prio.

**Comment by warpech** [comment]
All helper projects are merged to main app projects in the PR #313.

**Comment by warpech** [comment]
Done in #313
